Added unty dependency and added type checks

VictorKoenders opened the pull request.

VictorKoenders committed:
Added unty dependency and added type checks

Commit:
Bumped unty 0.0.2

dtolnay commented on src/de/impls.rs:
```diff
@@ -446,28 +446,26 @@ where
+            decoder.reader().read(&mut buf)?;
+            let ptr = &mut buf as *mut _ as *mut [T; N];
+
+            // Safety: we know that T is a u8, so it is perfectly safe to
```
The documentation of `unty` does not support this statement, but I think it's a documentation bug in unty. The documentation for `type_equal` says:

> _"Note that this may give false positives if any of the types have lifetimes."_

So, since `T` might contain lifetimes here, according to the documentation "T is u8" might be a false positive, so `T` might not be `u8` here.

Probably it needs to say the false positive is when both of the types have lifetimes.

VictorKoenders replied:
Updated documentation in unty 0.0.3, is this good now?

dtolnay replied:
:+1:

Consider also updating `impl Encode for [T; N]`, and the 3 `VecDeque` impls.

Commit:
Bump unty to 0.0.3

Commit:
Removed unneeded + Sized requirements
Optimize encode for [T; N]
Made BinaryHeap<T> proxy to Vec<T>
Made VecDeque decode/borrowdecode proxy to Vec<T>
Optimize VecDeque::<u8>::Encode to write 2 slices directly
Optimize Vec<u8> borrowdecode implementation

VictorKoenders commented:
trunk:
```
bench v1                time:   [32.759 µs 32.773 µs 32.789 µs]
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe

bench v2 (standard)     time:   [37.312 µs 37.329 µs 37.344 µs]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe

bench v2 (legacy)       time:   [31.839 µs 31.847 µs 31.855 µs]
Found 13 outliers among 100 measurements (13.00%)
  9 (9.00%) high mild
  4 (4.00%) high severe

bench v1 decode         time:   [197.19 µs 197.24 µs 197.28 µs]
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high severe

Benchmarking bench v2 decode (legacy): Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.7s, enable flat sampling, or reduce sample count to 60.
bench v2 decode (legacy)
                        time:   [1.1283 ms 1.1285 ms 1.1287 ms]
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  8 (8.00%) high severe
```

this PR:
```
bench v1                time:   [33.078 µs 33.084 µs 33.091 µs]
                        change: [+0.7552% +0.8417% +0.9214%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high severe

bench v2 (standard)     time:   [37.329 µs 37.336 µs 37.344 µs]
                        change: [+0.0325% +0.1377% +0.2356%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  4 (4.00%) high severe

bench v2 (legacy)       time:   [32.700 µs 32.704 µs 32.708 µs]
                        change: [+2.5681% +2.6751% +2.7748%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) high mild
  3 (3.00%) high severe

bench v1 decode         time:   [190.66 µs 190.70 µs 190.74 µs]
                        change: [-3.3484% -3.2612% -3.1718%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
  2 (2.00%) high severe

bench v2 decode (legacy)
                        time:   [216.64 µs 216.67 µs 216.70 µs]
                        change: [-80.807% -80.793% -80.778%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  6 (6.00%) high mild
  4 (4.00%) high severe
```

or tldr

||bincode 1|bincode 2|
|---|---|---|
|encode| 33.084 µs |32.704 µs|
|decode|190.70 µs|216.67 µs|
